src/app/shared/operators/build-up.ts

```diff
@@ -0,0 +1,26 @@
+
+import { OperatorFunction, pipe, concat, of, Observable } from 'rxjs';
+import { switchMap, map } from 'rxjs/operators';
+
+interface ObjectT { [k: string]: any }
+
+/**
+ * RxOperator which, for each input, first emits the raw input and then emits the raw input extended with the emission of an another
+ * observable based on this input.
+ * It is intended to be used to build up data from several async services. So it emits first the input (typically the output from another
+ * service) and then uses this input for another call (e.g. request more now we know this initial info) and finally emits the merge of
+ * the initial input with the output of the other call.
+ * For instance, it emits `{ name: 'foo', childId: 4 }` and then `{ name: 'foo', childId: 4, childName: 'bar' }` when the child name has
+ * been fetched.
+ * Note that this operator uses `switchMap`, so if the source re-emits, the subrequest is cancelled.
+ */
+export function buildUp<T extends ObjectT, U extends ObjectT>(project: (value: T) => Observable<U>): OperatorFunction<T, T | (T & U)> {
+  return pipe(
+    switchMap(initialValue => concat(
+      of(initialValue),
+      project(initialValue).pipe(
+        map(additionalValue => ({ ...initialValue, ...additionalValue }))
+      ),
+    ))
+  );
+}
```

Comment:
I understand the interest of enriching/projecting new data on the source observable data, but I don't understand the interest of emitting the initial value(s)
To me it adds complexity to the operator directly by implying a union type and potential extra computations

For instance if you `buildUp()` more than twice, let's say 5 times, the output type will be unbearable to understand and eventually to check

Why is emitting the initial value(s) needed ?

Reply:
That's all the point of the operator 😁, to hide this messy part of emitting the initial value. 
Actually as you saw the example in get-item-by-id and get-group-by-id, you might get the point.
So for items, we want 4 info: the breadcrumb, the item itself, the results and the nav data.
We first fetch (in parallel) breadcrumb and the item itself as "results" needs info from the previous ones, and nav data needs results. However, we do not want to wait for the 4 requests to complete (3 of them in cascade) before displaying anything, most of the interesting information is already there after the first set of 2 requests. So we emit these 2 values, use them in the interface, and in parallel the whole set of data is "build up" on top of that...
So we need 2 emissions for build up, the first one immediately with the data, then the other one, asynced, based on this data.

Reply:
I still doubt the idea because emitting new values of the state will trigger unneeded re-renders
To me, either the states are deeply coupled and only usable as a whole, either they should be split into different smaller states which you can subscribe to separately

Knowing that and since this subject is only a matter of opinion, if it is still what you want then fine by me

Reply:
I still doubt the idea because emitting new values of the state will trigger unneeded re-renders
To me, either the states are deeply coupled and only usable as a whole, either they should be split into different smaller states which you can subscribe to separately

Knowing that and since this subject is only a matter of opinion, if it is still what you want then fine by me

Reply:
Conversation to resolve after reading